design-docs/improved-console.md

```diff
@@ -0,0 +1,422 @@
+
+#### DC2 - Logging consumed through Tooling API/GUI/CI
+Consider the case where rich text is not available, but there are other means to provide rich output. How would IntelliJ, a web page, or some other GUI behave/display given output.
+ 
+#### DC3 - Non-interactive output
+Consider how console output should behave when an interactive terminal is not available. For example, when output is piped to some other process or a log file. 
+
+#### DC4 - Extensible for future features of Gradle
+Consider how build scans, task output cache, and other yet undreamt features are represented. 
+
+#### DC5 - Performance
+All of the above should be achieved with nominal performance impact.
+
+#### DC6 - Backwards Compatibility
+Consider tools that parse the current output for their own highlighting. Provide a fallback to do no harm in order to avoid blocking tool providers upgrading. Consider different consoles and OSes. `System.in`, `--continuous`
+
+Consider: 
+ - Different terminals
+ - Windows, Linux, macOS
+ - `System.in`
+ - `TERM=dumb`
+ - Fonts with limited provided characters outside ASCII
+
+## Story: Display parallel work in-progress independently by operation
+Show incomplete ProgressOperations on separate lines, up to a specified maximum number. 
+
+This is opt-in initially by specifying the `org.gradle.console.parallel=true` Gradle property.
+
+### User Visible Changes
+The mainArea is displayed at the top for rendering important messages like warnings
+and errors.
+
+`TaskExecutionLogger` no longer sets logging headers. This prevents logging of task names
+in main area. To compensate for the user's loss of context, [WARN and ERROR logs can be 
+configured to give more context](#story-log-messages-can-be-configured-to-give-more-context),
+possibly setting the log category (and allowing `LogEvent` to accept a category)
+to the build operation description then configuring `StyledTextOutputBackedRenderer`
+to render category by default.
+
+In the case where a build process is not attached to a terminal, `OutputEvent`s are
+handled the same as previous versions of Gradle in that they are streamed to
+stdout and stderr with no filtering or throttling.
+ 
+The `ProgressLabel` is intended to give the user a glance-able indicator that tells
+them whether the build will be finished soon or not. 
+See ["Display build progress as a progress bar through colorized output"](#story-display-build-progress-as-a-progress-bar-through-colorized-output) below.
+
+The multiple progressArea lines are rendered results of a fixed list of `AbstractWorker`s. 
+It displays work-in-progress for each `AbstractWorker`. `BuildOperationDetails` is built
+with an optional worker ID to provide information about "where" an operation is running
+so that a mapping of worker to index of operation status line can be maintained.
+
+A `OperationStatusBarFormatter` can be used to customize the intra-line display of each 
```

Comment:
We currently have a `DefaultStatusBarFormatter` that adds the `>`s — this is very similar to that and doesn't have any user facing implications. In later stories, this might become configurable, but it's optional.